scipy/memetic_climbing.py

```diff
@@ -0,0 +1,253 @@
+    minimum of an objective function. Optionally, it performs a local search
+    using L-BFGS-B to refine the solution.
+
+    Parameters
+    ----------
+    objective : callable
+        The objective function to be minimized. It should accept different arrays of dimensions.
+    bounds : sequence of tuples
+        Bounds for the variables. Each tuple represents the (min, max) bounds
+        for a single dimension.
+    nx : int
+        Number of dimensions for the input space.
+    cycles : int, optional
+        Number of cycles to perform (default is 1000).
+    max_iterations : int, optional
+        Maximum number of iterations (default is 20000).
+    step_size : float, optional
+        Step size for hill climbing (default is 20).
+    decay_rate : float, optional
+        Decay rate for step size (default is 0.99977).
+    local_search : bool, optional
+        Whether to perform local search using L-BFGS-B (default is True).
+
+    Returns
+    -------
+    iteration_details_global : list
+        List of details for each iteration.
+    best_result : dict
+        Dictionary containing the best found value, with the key:
+        - 'best_value': The minimum value found for the objective function.
+
+
+    Notes
+    -----
+    The hill climbing process starts with random points and adjusts them according
+    to the specified step size. If `local_search` is enabled, a final optimization
+    using L-BFGS-B is performed on the best found solution.
+    """
+    best_global_value = float('inf')
+    iteration_details_global = []
+    max_iterations = max_iterations // cycles
+
+    lower_bounds, upper_bounds = np.array([b[0] for b in bounds]), np.array([b[1] for b in bounds])
+
+    for cycle in range(cycles):
+
+        x0 = np.random.uniform(lower_bounds, upper_bounds, size=nx)
+        best_x = np.array(x0)
+
+        best_value = objective(best_x)
+
+        iterations = 0
+
+        while iterations < max_iterations:
+            next_x = best_x + hollow_distribution(nx) * step_size
+            next_x = np.clip(next_x, a_min=lower_bounds, a_max=upper_bounds)
+            next_value = objective(next_x)
+
+            if next_value <= best_value:
+                best_x, best_value = next_x, next_value
+
+            iterations += 1
+            step_size *= decay_rate
+
+        if local_search:
+            result_bfgs = minimize(objective, best_x, bounds=bounds, method='L-BFGS-B')
+            best_x_bfgs, best_value_bfgs = result_bfgs.x, result_bfgs.fun
+
+            if best_value_bfgs <= best_value:
+                best_x, best_value = best_x_bfgs, best_value_bfgs
+
+        if best_value < best_global_value:
+            best_global_value = best_value
+
+    return iteration_details_global, {
+        # 'best_x': best_global_x,
+        'best_value': best_global_value,
+    }
+
+def memetic_climbing(objective, bounds, nx, cycles=1000, local_search=True):
```

Comment:
This function should use exactly the same nomenclature as `scipy.optimize`. e.g. `func` instead of `objective`. `maxiter` instead of `cycles`, etc.

Reply:
`bounds` should be either `(min, max)` tuples or a `Bounds` instance.

Reply:
Part of the minimisation is stochastic, the function should have some way of passing in a `np.random.Generator`, or a seed, to ensure repeatable minimisation.